### PR TITLE
[RFR] Add Error Page

### PR DESCRIPTION
**Pull request by fzaninotto**
- [x] Create error boundary page
- [x] Find a way to make navigation work after an error
- [x] Document how to override it

kapture 2018-07-05 at 10 29 54

### CODE REVIEW

File: packages/ra-ui-materialui/src/layout/Error.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        <Hidden mdUp>
+            <AppBarMobile />
+        </Hidden>
+        <div className={classnames(classes.container, className)} {...rest}>
+            <h1 className={classes.title}>
+                <ErrorIcon className={classes.icon} />
+                {translate('ra.page.error')}
+            </h1>
+            <div>{translate('ra.message.error')}</div>
+            <ExpansionPanel className={classes.panel}>
```
Comment (djhi) [inline packages/ra-ui-materialui/src/layout/Error.js]: Might be a good idea to only display this panel when `process.env.NODE_ENV` is `development` ?

Comment (fzaninotto) [reply packages/ra-ui-materialui/src/layout/Error.js]: you're right.

File: packages/ra-ui-materialui/src/layout/Layout.js [hunk 2]
```diff
@@ -53,18 +54,35 @@ const styles = theme => ({
...
-            errorInfo,
+    constructor(props) {
+        super(props);
+        /**
+         * Reset the error state upon navigation
+         *
+         * @see https://stackoverflow.com/questions/48121750/browser-navigation-broken-by-use-of-react-error-boundaries
+         * */
+        props.history.listen(() => {
+            if (this.state.hasError) {
```
Comment (djhi) [inline packages/ra-ui-materialui/src/layout/Layout.js]: neat :+1:

File: packages/ra-ui-materialui/src/layout/Error.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    className,
+    translate,
+    ...rest
+}) => (
+    <Fragment>
+        <Hidden mdUp>
+            <AppBarMobile />
+        </Hidden>
+        <div className={classnames(classes.container, className)} {...rest}>
+            <h1 className={classes.title}>
```
Comment (djhi) [inline packages/ra-ui-materialui/src/layout/Error.js]: Accessibility: add the `role` attribute with the `alert` value.

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_alert_role

**Pull request merged**